File: backend/internal/models/mythicplus/mythicplusruns.go
package models

import (
	"time"
)

// MythicPlusRun represents a run in the Mythic+ dungeon
// It includes the seasons, the dungeon, the affixes, the members, the score, the duration, etc.
type MythicPlusRun struct {
	CompletedTimestamp    time.Time
	DungeonID             uint
	Dungeon               Dungeon
	ShortName             string
	Duration              int64
	IsCompletedWithinTime bool
	KeyStoneUpgrades      int
	KeystoneLevel         int
	MythicRating          float64
	SeasonID              uint
	Season                Season
	Affixes               []Affix
	Members               []MythicPlusRunMember
}

// MythicPlusRunMember represents a member in the Mythic+ dungeon run
type MythicPlusRunMember struct {
	CharacterID       uint
	CharacterName     string
	RealmID           uint
	RealmName         string `json:"-"`
	RealmSlug         string
	EquippedItemLevel int
	RaceID            uint
	RaceName          string
	SpecializationID  uint
	Specialization    string
}

// MythicPlusRunAffix represents an affix in the Mythic+ dungeon
type MythicPlusRunAffix struct {
	MythicPlusRunID uint
	AffixID         uint
}

File: frontend/src/types/mythicPlusRuns.ts
export interface Season {
  slug: string;
  name: string;
  shortName: string;
  id: number;
  Dungeons: null | Dungeon[];
}

export interface Dungeon {
  ID: number;
  ChallengeModeID: number;
  EncounterID: number;
  Slug: string;
  Name: string;
  ShortName: string;
  MediaURL: string;
  Icon: string;
  KeyStoneUpgrades: KeyStoneUpgrade[];
  Seasons: Season[];
}

export interface Affix {
  ID: number;
  Name: string;
  Icon: string;
  WowheadURL: string;
}

export interface Member {
  CharacterID: number;
  CharacterName: string;
  RealmID: number;
  RealmSlug: string;
  EquippedItemLevel: number;
  RaceID: number;
  RaceName: string;
  SpecializationID: number;
  Specialization: string;
}

export interface MythicPlusSeasonInfo {
  CharacterName: string;
  RealmSlug: string;
  SeasonID: number;
  OverallMythicRating: number;
  OverallMythicRatingHex: string;
  BestRuns: MythicPlusRuns[];
}

export interface MythicPlusRuns {
  CompletedTimestamp: string;
  DungeonID: number;
  Dungeon: Dungeon;
  ShortName: string;
  Duration: number;
  IsCompletedWithinTime: boolean;
  KeyStoneUpgrades: number;
  KeystoneLevel: number;
  MythicRating: number;
  SeasonID: number;
  Season: Season;
  Affixes: Affix[];
  Members: Member[];
}

export interface KeyStoneUpgrade {
  ChallengeModeID: number;
  QualifyingDuration: number;
  UpgradeLevel: number;
}

export interface MythicDungeonProps {
  characterName: string;
  realmSlug: string;
  namespace: string;
  locale: string;
  region: string;
  seasonSlug: string;
}
